File: api2/controller/SwitchController.php
<?php

require_once './core/Request.php';
require_once './core/Response.php';
require_once './model/SwitchModel.php';
require_once './model/MacModel.php';
require_once './model/UserModel.php';

class SwitchController
{
    
    public function createSwitch(Request $request, Response $response)
    {
        try {
            
            $data            = $request->bodyJson();
            $mac_address     = $data['mac_address'];
            $user            = UserModel::getUserByUUID($data['user_id']);
            $data['user_id'] = $user['id'];
            //$macAvailable    = MacModel::checkMacAddressAvailable($mac_address);
            $switchExists    = SwitchModel::checkSwitchExists($data);
            
            
            if($switchExists){
                $data = SwitchModel::reactivateSwitch($data);
                
                if ($data) :
                    $response::json([
                        'status' => 'success',
                        'dados' => $data
                    ], 200);
                    else :
                        $response::json([
                            'status' => 'error',
                            'msg' => 'Internal Error'
                        ], 400);
                    endif;
                }else {
                //if ($macAvailable){
                   
                    
                    $reslt = SwitchModel::createSwitch($data);
                    
                    if ($reslt) :
                       // MacModel::updateMacAddressRecord($mac_address);
                        $response::json([
                            'status' => 'success',
                            'dados' => $reslt
                        ], 200);
                        else :
                            $response::json([
                                'status' => 'error',
                                'msg' => 'Internal Error'
                            ], 400);
                        endif;
                        
                    }
                    // else{
                    //     $response::json([
                    //         'status' => 'error',
                    //         'msg' => 'Invalid Mac Address'
                    //     ], 401);
                    // }
                    
                } catch (\Exception $e) {
                    $response::json([
                        'status' => 'error',
                        'msg' => $e->getMessage()
                    ], 500);
                }
    }
            
    public function updateSwitch(Request $request, Response $response)
    {
        try {
            
            $data = $request->bodyJson();
            
            $data = SwitchModel::updateSwitchInfo($data);
            
            if ($data) {
                $response::json([
                    'status' => 'success',
                    'dados' => $data
                ], 200);
            }else {
                $response::json([
                    'status' => 'error',
                    'msg' => 'Internal Error'
                ], 400);
            }
            
            
        } catch (\Exception $e) {
            $response::json([
                'status' => 'error',
                'msg' => $e->getMessage()
            ], 500);
        }
    }


    public function deleteSwitch(Request $request, Response $response)
    {
        try {
            
            $data = $request->bodyJson();
            
            $data = SwitchModel::softDelete($data['mac_address']);
            
            if ($data) {
                $response::json([
                    'status' => 'success',
                    'dados' => $data
                ], 200);
            }else {
                $response::json([
                    'status' => 'error',
                    'msg' => 'Internal Error'
                ], 400);
            }
            
            
        } catch (\Exception $e) {
            $response::json([
                'status' => 'error',
                'msg' => $e->getMessage()
            ], 500);
        }
    }

    public function toggleSwitch(Request $request, Response $response)
    {
        try {
            
            $data = $request->bodyJson();
            
            $data = SwitchModel::toggleSwitch($data);
            
            if ($data) {
                $response::json([
                    'status' => 'success',
                    'dados' => $data
                ], 200);
            }else {
                $response::json([
                    'status' => 'error',
                    'msg' => 'Internal Error'
                ], 400);
            }
            
            
        } catch (\Exception $e) {
            $response::json([
                'status' => 'error',
                'msg' => $e->getMessage()
            ], 500);
        }
    }

    public function getSwitches(Request $request, Response $response)
    {
        try {
            $data = $request->bodyJson();
            $userId = $data['user_id'];
            
            $switches = SwitchModel::getSwitches($userId);
            
            if ($switches) {
                return $response::json([
                    'status' => 'success',
                    'data' => $switches
                ], 200);
            } else {
                return $response::json([
                    'status' => 'error',
                    'message' => 'Nenhum ponto encontrado para o usuário',
                ], 404);
            }
        } catch (\Exception $e) {
            return $response::json([
                'status' => 'error',
                'message' => $e->getMessage()
            ], 500);
        }
    }

    public function getSwitch(Request $request, Response $response)
    {
        try {
            $data = $request->bodyJson();
            $data = SwitchModel::getSwitch($data['mac_address']);
            
            if ($data) {
                $response::json([
                    'status' => 'success',
                    'dados' => $data
                ], 200);
            }else {
                $response::json([
                    'status' => 'error',
                    'msg' => 'Ponto não encontrado'
                ], 404);
            }
        } catch (\Exception $e) {
            $response::json([
                'status' => 'error',
                'msg' => $e->getMessage()
            ], 500);
        }
    }
}
        

File: lib/controller/switch_controller.dart
import 'package:flutter/material.dart';
import '../model/switch_model.dart';
import '../provider/switch_provider.dart';
import '../utils/auth_provider.dart';

class SwitchController extends ChangeNotifier {
  final SwitchProvider provider;
  final String userId = AuthProvider().userId!;

  SwitchController({required this.provider});

  Future<bool> createSwitch(
      String name, String watts, String macAddress) async {
    try {
      final result = await provider.createSwitch(
        name,
        watts,
        macAddress,
        userId,
      );
      notifyListeners();

      return result['status'] == 'success';
    } catch (error) {
      return false;
    }
  }

  Future<List<SwitchModel>> getSwitches() async {
    try {
      final switches = await provider.getSwitches(userId);
      notifyListeners();
      return switches;
    } catch (error) {
      return [];
    }
  }

  Future<List<SwitchModel>> getSwitchesWithoutGroup() async {
    try {
      final switches = await provider.getSwitchesWithoutGroup(userId);
      notifyListeners();
      return switches;
    } catch (error) {
      return [];
    }
  }

  Future<bool> updateSwitch({
    required String name,
    required String watts,
    required String macAddress,
  }) async {
    try {
      final Map<String, dynamic> data = {
        'name': name,
        'watts': watts,
        'mac_address': macAddress,
      };

      final success = await provider.updateSwitch(data);
      if (success) {
        notifyListeners();
      }
      return success;
    } catch (error) {
      return false;
    }
  }

  Future<bool> deleteSwitch({required String macAddress}) async {
    try {
      final result = await provider.deleteSwitch(macAddress);
      notifyListeners();
      return result['status'] == 'success';
    } catch (error) {
      return false;
    }
  }

  Future<bool> toggleSwitch(
      {required String macAddress, required bool isActive}) async {
    try {
      final Map<String, dynamic> data = {
        'mac_address': macAddress,
        'user_id': userId,
        'is_active': isActive,
      };
      final success = await provider.toggleSwitch(data);
      notifyListeners();
      return success['status'] == 'success';
    } catch (error) {
      return false;
    }
  }
}
